src/rust/engine/src/nodes.rs

```diff
@@ -1146,8 +1146,9 @@ impl Task {
                 // `type(input) != input_type`.
                 throw(format!(
                   "{get} was not detected in your @rule body at rule compile time. \
-                  Was the `Get` constructor called in a separate function, or perhaps \
-                  dynamically? If so, it must be inlined into the @rule body.",
+                  Was the `Get` constructor called in a non async-function, or \
+                  was it inside an async function defined after the @rule? \
+                  Make sure the `Get` is defined before or inside the @rule body.",
```

Comment:
Is this truly accurate?

It should be OK to have `Get` constructors in async functions too now. If not, we'd like to learn about it so we can identify why that is.

I mean, the old phrasing is stale now too.. so a change is due, just want it to be accurate :)

Reply:
>It should be OK to have Get constructors in async functions too now.

The first sentence is asking about potential incorrect usages, calling `Get` in a non-async function will fail, and so will calling a rule helper containing `Get` _before_ its definition (i.e rule helper has to be above rule).

I'm open for suggested changes in wording, I tbh dont know all error modes, I just think the current message isn't really acceptable :sweat_smile:

Reply:
To be clear: a5c7905f974e187f13ca1e6a7db11b66d8f78200 this was the issue I did not understand to be a problem until stu pointed it out.

Reply:
this is strange. I'll need to take a closer look at what is going on here, as I copied your test to the rule visitor (which does the rule parsing at run time that is mentioned in the error message) and it picks this up just fine:
```diff
git diff
diff --git a/src/python/pants/engine/internals/rule_visitor_test.py b/src/python/pants/engine/internals/rule_visitor_test.py
index 4444ac3706..d89394db4d 100644
--- a/src/python/pants/engine/internals/rule_visitor_test.py
+++ b/src/python/pants/engine/internals/rule_visitor_test.py
@@ -321,3 +321,16 @@ def test_union_types() -> None:
         Get(str, mc.b)

     assert_awaitables(somerule, [(str, bool)])
+
+
+#@rule
+async def uses_rule_helper_before_definition() -> int:
+    return await get_after_rule()
+
+
+async def get_after_rule() -> int:
+    return await Get(int, str, "hello")
+
+
+def test_outlined() -> None:
+    assert_awaitables(uses_rule_helper_before_definition, [(int, str)])
```

Reply:
Don't you need to move the test above the rule & helper to mimic?

Reply:
That worked too, but I guess that is due to the file being fully parsed before the tests run, where as when using the `@rule` decorator we start processing at the time the rule is declared mid-import. So these rule visitor tests will never catch that.

Ah, which makes me realize the underlying issue, being just that (that we start processing the rule body before the entire file has been parsed)